src/librustc_mir/interpret/memory.rs

```diff
@@ -992,12 +992,12 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> HasMemory<'a, 'mir, 'tcx, M> for Me
 impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> HasMemory<'a, 'mir, 'tcx, M> for EvalContext<'a, 'mir, 'tcx, M> {
     #[inline]
     fn memory_mut(&mut self) -> &mut Memory<'a, 'mir, 'tcx, M> {
-        &mut self.memory
+        self.memory_mut()
```

Comment:
This feels very infinite recursion dangerous. Or at least surprising to read. Can we just get rid of the `memory_mut` and `memory` inherent methods in favour of this trait impl?

Reply:
`HasMemory` is private to this crate, and `tools/miri` would need to call these methods.

The solution above would obviate this problem.